README.md
# MJCast

Streaming MJPEG screenshots.

We created this project for streaming monitoring with Raspberry.
Raspberry was slow when we loaded web pages with many graphics,
so we decide to display screenshots with MJPEG procotol.

Installation on Debian
======================

- apt-get install chromium-driver chromium python-dev python-pip redis-server
- pip install -r requirements.txt
- pip install mjcast

Running with Docker
===================

If you want to run this project with docker you first need to build it.

```sh
docker build -t mjcast .
```

Then you would run it with the following command:
```sh
docker run -d -p 8670:8670 mjcast
```

Results
=======

Streaming URL: http://localhost:8670/server/stream


debian/changelog
mjcast (0.1.12) unstable; urgency=medium

  * Reviewed requirements.

 -- Adrien DELLE CAVE (Decryptus) <adc@doowan.net>  Sat, 23 Feb 2019 07:59:56 +0100

mjcast (0.1.11) unstable; urgency=medium

  * Reviewed requirements.

 -- Adrien DELLE CAVE (Decryptus) <adc@doowan.net>  Fri, 22 Feb 2019 19:31:05 +0100

mjcast (0.1.10) unstable; urgency=medium

  * Reviewed requirements.

 -- Adrien DELLE CAVE (Decryptus) <adc@doowan.net>  Fri, 22 Feb 2019 19:01:36 +0100

mjcast (0.1.9) unstable; urgency=medium

  * Release version: 0.1.9

 -- Adrien DELLE CAVE (Decryptus) <adc@doowan.net>  Fri, 05 Oct 2018 03:12:45 +0200

mjcast (0.1.3) unstable; urgency=medium

  * Fixed requirements.txt.

 -- Adrien DELLE CAVE (Decryptus) <adc@doowan.net>  Mon, 25 Jun 2018 11:46:33 +0200

mjcast (0.1.2) unstable; urgency=medium

  * Fixed bin/mjcast and reviewed requirements.txt.

 -- Adrien DELLE CAVE (Decryptus) <adc@doowan.net>  Mon, 25 Jun 2018 11:26:38 +0200

mjcast (0.1.1) unstable; urgency=medium

  * Fixed WebDriverException: Message: no such session.

 -- Adrien DELLE CAVE (Decryptus) <adc@doowan.net>  Wed, 09 May 2018 11:13:42 +0200

mjcast (0.1.0) unstable; urgency=medium

  * Initial release.

 -- Adrien DELLE CAVE (Decryptus) <adc@doowan.net>  Wed, 04 Apr 2018 11:55:08 +0200


mjcast/plugins/playlist.py
# -*- coding: utf-8 -*-
"""mjcast channels"""

__author__  = "Adrien DELLE CAVE <adc@doowan.net>"
__license__ = """
    Copyright (C) 2018  doowan

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA..
"""

import base64
import cv2
import os
import logging
import threading
import time

from datetime import datetime
from dwho.adapters.redis import DWhoAdapterRedis
from dwho.classes.plugins import DWhoPluginBase, PLUGINS
from selenium import webdriver
from selenium.common.exceptions import TimeoutException, WebDriverException
from selenium.webdriver.chrome.options import Options
from sonicprobe.libs.keystore import Keystore
from tempfile import mkstemp


LOG              = logging.getLogger('mjcast.playlist')

DATETIME_DEFAULT = {'position': 'auto',
                    'format': "%Y-%m-%d %H:%M:%S %Z",
                    'face': cv2.FONT_HERSHEY_PLAIN,
                    'color': [0, 0, 0],
                    'scale': 0,
                    'thickness': 1}


class MjCastMedia(threading.Thread):
    def __init__(self, name, params, adapter_redis):
        threading.Thread.__init__(self)
        self.daemon        = True
        self.killed        = False
        self.result        = Keystore().add(self.name)

        self.name          = name
        self.params        = params
        self.adapter_redis = adapter_redis
        self.ttl           = int(self.params.get('ttl') or 1)

        self.init_driver()

    def get_name(self):
        return self.name

    def get_result(self):
        return self.result

    def add_datetime(self, img):
        if not isinstance(self.params.get('datetime'), dict):
            return img

        dt = DATETIME_DEFAULT.copy()
        dt.update(self.params['datetime'])

        text      = time.strftime(dt['format'])
        textsize  = cv2.getTextSize(text, dt['face'], float(dt['scale']), dt['thickness'])

        if dt['position'] == 'auto':
            dt['org'] = (img.shape[1] - (textsize[0][0] + 10), 20)

        cv2.putText(img,
                    text,
                    dt['org'],
                    dt['face'],
                    float(dt['scale']),
                    dt['color'],
                    dt['thickness'])

        return img

    def save_screenshot(self, with_delay = True):
        if with_delay and self.params.get('delay'):
            time.sleep(float(self.params['delay']))

        (tmpfd, tmpfile, filepath, data) = (0, None, None, None)

        try:
            (tmpfd, tmpfile) = mkstemp()
            filepath = "%s.png" % tmpfile
            os.rename(tmpfile, filepath)
            self.driver.save_screenshot(filepath)
            img      = self.add_datetime(cv2.imread(filepath))

            gmtime   = datetime.utcnow().strftime('%s')
            data     = cv2.imencode('.jpg', img)[1].tostring()
            xlen     = len(data)

            self.adapter_redis.set_key(self.name,
                                       "%s|%d|%s" % (gmtime,
                                                     xlen,
                                                     base64.b64encode(data)),
                                       self.ttl)

            return (gmtime, xlen, data)
        except (TimeoutException, WebDriverException):
            self.reinit_driver()
            self.save_screenshot(with_delay = with_delay)
        finally:
            data = None

            if tmpfd:
                os.close(tmpfd)

            if tmpfile and os.path.exists(tmpfile):
                os.unlink(tmpfile)

            if filepath and os.path.exists(filepath):
                os.unlink(filepath)

    def init_driver(self):
        self.driver = webdriver.Chrome(
                          executable_path = self.params['executable_path'],
                          chrome_options  = self.driver_options())

    def reinit_driver(self):
        self.quit()
        self.init_driver()
        self.driver.get(self.params['url'])

    def driver_options(self):
        r = Options()

        for option in self.params['options']:
            r.add_argument(option)

        r.binary_location = self.params['binary_location']

        return r

    def run(self):
        self.driver.get(self.params['url'])
        self.save_screenshot()

        while not self.killed:
            try:
                value = self.adapter_redis.get_key(self.name)
                wait  = float(self.params.get('wait', 1))
                if value:
                    (gmtime, xlen, data) = value.split('|', 2)
                    self.result.try_acquire(self.name)
                    (self.result.set(self.name, 'gmtime', gmtime)
                                .set(self.name, 'length', long(xlen))
                                .set(self.name, 'data', base64.b64decode(data))
                                .set(self.name, 'wait', wait))
                    self.result.try_release(self.name)
                else:
                    with_delay = False
                    if bool(self.params.get('refresh', True)):
                        with_delay = True
                        try:
                            self.driver.refresh()
                        except (TimeoutException, WebDriverException):
                            self.reinit_driver()

                    (gmtime, xlen, data) = self.save_screenshot(with_delay)
                    self.result.try_acquire(self.name)
                    (self.result.set(self.name, 'gmtime', gmtime)
                                .set(self.name, 'length', long(xlen))
                                .set(self.name, 'data', data)
                                .set(self.name, 'wait', wait))
                    self.result.try_release(self.name)
                time.sleep(self.ttl)
            except Exception, e:
                LOG.exception(repr(e))
            finally:
                self.result.try_release(self.name)

    def quit(self):
        try:
            self.driver.quit()
        except (TimeoutException, WebDriverException):
            pass

    def terminate(self):
        self.killed = True

        self.quit()

        self.driver = None


class MjCastPlaylistPlugin(DWhoPluginBase):
    PLUGIN_NAME = 'playlist'

    def safe_init(self):
        self.medias        = []
        self.adapter_redis = DWhoAdapterRedis(self.config, prefix = 'playlist')

        for params in self.config['plugins']['playlist']:
            self.medias.append(MjCastMedia(params['name'], params, self.adapter_redis))

    def get_medias(self):
        return self.medias

    def at_start(self):
        for media in self.medias:
            media.start()

    def at_stop(self):
        for media in self.medias:
            media.terminate()


if __name__ != "__main__":
    def _start():
        PLUGINS.register(MjCastPlaylistPlugin())
    _start()


mjcast/modules/server.py
# -*- coding: utf-8 -*-
"""server module"""

__author__  = "Adrien DELLE CAVE"
__license__ = """
    Copyright (C) 2018  doowan

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA..
"""

import logging
import time

from dwho.classes.modules import DWhoModuleBase, MODULES
from dwho.classes.plugins import PLUGINS

LOG            = logging.getLogger('mjcast.modules.server')

_BOUNDARY_NAME = 'frame'
_CONTENT_TYPE  = "multipart/x-mixed-replace; boundary=%s" % _BOUNDARY_NAME
_CACHE         = {'loading': None}


class MjCastServerModule(DWhoModuleBase):
    MODULE_NAME = 'server'

    def _push_loading(self, request):
        if not self.config['general'].get('loading_file'):
            return

        if not _CACHE['loading']:
            f = None

            with open(self.config['general']['loading_file'], 'rb') as f:
                data = f.read()
                _CACHE['loading'] = (len(data), data)
                data = None

            if f:
                f.close()

        LOG.info("streaming loading")

        request.end_headers()
        request.wfile.write("--" + _BOUNDARY_NAME)
        request.end_headers()
        request.send_header('Content-Type', 'image/jpeg')
        request.send_header('Content-Length', _CACHE['loading'][0])
        request.end_headers()
        request.wfile.write(_CACHE['loading'][1])

    def status(self, request):
        return True

    def stream(self, request):
        request.send_response(200)
        request.send_header('Content-type', _CONTENT_TYPE)
        request.send_header('Cache-Control', 'no-store, no-cache, must-revalidate, pre-check=0, post-check=0, max-age=0')
        request.send_header('Pragma', 'no-cache')
        request.end_headers()

        while True:
            for media in PLUGINS['playlist'].get_medias():
                while True:
                    result = media.get_result()
                    if not result:
                        self._push_loading(request)
                        time.sleep(0.5)
                        continue

                    try:
                        result.try_acquire(media.get_name())
                        gmtime = result.get(media.get_name(), 'gmtime')
                        xlen   = result.get(media.get_name(), 'length')
                        data   = result.get(media.get_name(), 'data')
                        wait   = result.get(media.get_name(), 'wait')
                    except Exception, e:
                        LOG.error(e)
                    finally:
                        result.try_release(media.get_name())

                    if data:
                        break

                    self._push_loading(request)
                    time.sleep(0.5)

                LOG.info("streaming %r ready", media.get_name())
                begin = time.time()

                while begin + wait >= time.time():
                    request.end_headers()
                    request.wfile.write("--" + _BOUNDARY_NAME)
                    request.end_headers()
                    request.send_header('X-Mjcast-Name', media.get_name())
                    request.send_header('X-Mjcast-Gmtime', gmtime)
                    request.send_header('X-Mjcast-Wait', wait)
                    request.send_header('Content-Type', 'image/jpeg')
                    request.send_header('Content-Length', xlen)
                    request.end_headers()
                    request.wfile.write(data)
                    if ((begin + wait) - time.time()) >= 0.5:
                        time.sleep(0.5)


if __name__ != "__main__":
    def _start():
        MODULES.register(MjCastServerModule())
    _start()
